### ISSUE TITLE
Exception when updating a Buffer type to null

### ISSUE THREAD

**Issue by ValYouW** [Automattic/mongoose#4944]
Hi,

I have a simple schema with a Buffer property, when creating a document I can set the property to `null` with no issues, but when issuing an update command setting the property to `null` I get an exception, is it not a valid operation?

Example:
```js
var mongoose = require('mongoose');
var conn = mongoose.createConnection('mongodb://127.0.0.1/test', {user: 'test', pass: 'test'});

var ItemSchema = new mongoose.Schema({
    owner: { type: mongoose.Schema.Types.Buffer },
    name: {type: String},
}, { collection: 'items' });

var ItemModel = conn.model('Item', ItemSchema);

var item = new ItemModel({owner: null, name: 'item1'});
item.save().then(res => {
    console.log('Saved with owner null -->', res)

    // Update to some non-null value is OK
    ItemModel.findOneAndUpdate({name: 'item1'}, {owner: new Buffer('Joe')}, {new: true}).then(res => {
        console.log('Updated to Joe -->', res);

        console.log('Updating back to null will throw');
        ItemModel.findOneAndUpdate({name: 'item1'}, {owner: null}, {new: true}).then(r => console.log(r)).catch(e => console.log(e));
    });
});
```

Output:
```
Saved with owner null --> { __v: 0,
  owner: null,
  name: 'item1',
  _id: 5891b9116c280c1df8f7c58b }
Updated to Joe --> { _id: 5891b9116c280c1df8f7c58b,
  owner:
   Binary {
     _bsontype: 'Binary',
     sub_type: 0,
     position: 3,
     buffer: <Buffer 4a 6f 65> },
  name: 'item1',
  __v: 0 }
Updating back to null will throw
TypeError: Cannot read property 'toObject' of null
    at SchemaBuffer.castForQuery (C:\temp\mongo\node_modules\mongoose\lib\schema\buffer.js:181:24)
    at Query._castUpdateVal (C:\temp\mongo\node_modules\mongoose\lib\query.js:2881:17)
    at Query._walkUpdatePath (C:\temp\mongo\node_modules\mongoose\lib\query.js:2825:25)
    at Query._castUpdate (C:\temp\mongo\node_modules\mongoose\lib\query.js:2697:23)
    at castDoc (C:\temp\mongo\node_modules\mongoose\lib\query.js:2904:18)
    at Query._findAndModify (C:\temp\mongo\node_modules\mongoose\lib\query.js:1928:17)
    at Query._findOneAndUpdate (C:\temp\mongo\node_modules\mongoose\lib\query.js:1784:8)
    at C:\temp\mongo\node_modules\kareem\index.js:244:8
    at C:\temp\mongo\node_modules\kareem\index.js:18:7
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

Am I doing something wrong?
Looking at `SchemaBuffer.castForQuery@181`:
`return this.cast(val).toObject({ transform: false, virtuals: false });`
But `SchemaBuffer.cast` explicitly returns `null` at line 129, so I guess it should be handled in `SchemaBuffer.castForQuery@181`, simple fix:
```js
var casted = this.cast(val);
return casted ? casted.toObject({ transform: false, virtuals: false }) : casted;
```

Ideas? Thx!

Versions:
mongodb@3.4.0
mongoose@4.8.1
node@6.9.1

**Comment by ValYouW** [comment]
Fix idea:
https://github.com/ValYouW/mongoose/commit/6650dbddc0c8b365ccebdfa5fd5c60c2d9775f7b

**Comment by varunjayaraman** [comment]
Thanks for the quick repro script, looks like this is a bug

**Comment by ValYouW** [comment]
@varunjayaraman Can turn https://github.com/ValYouW/mongoose/commit/6650dbddc0c8b365ccebdfa5fd5c60c2d9775f7b into a PR if it helps...

**Comment by vkarpov15** [comment]
Fixed by eec2c362ab5bcb1bd412cdffc720d254702a4f13
